clash-ffi/src/Clash/FFI/VPI/Control.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Clash.FFI.VPI.Control
+#if defined(VERILOG_2001)
+  ( Control(..)
+  , StopValue(..)
+  , DiagnosticLevel(..)
+  , CouldNotControl(..)
+  , controlSimulator
+  ) where
+
+import           Control.Exception (Exception)
+import qualified Control.Monad as Monad (unless)
+import qualified Control.Monad.IO.Class as IO (liftIO)
+import           Foreign.C.Types (CInt(..))
+import           GHC.Stack (CallStack, callStack, prettyCallStack)
+
+import           Clash.FFI.Monad (SimCont)
+import qualified Clash.FFI.Monad as Sim (throw)
+
+data Control
+  = Stop DiagnosticLevel
+  | Finish DiagnosticLevel
+  | Reset StopValue (Maybe CInt) DiagnosticLevel
+  deriving stock (Show)
+
+data StopValue
+  = Interactive
+  | Batch
+  deriving stock (Show)
+
+stopValueToCInt :: StopValue -> CInt
+stopValueToCInt = \case
+  Interactive -> 0
+  Batch -> 1
+
+data DiagnosticLevel
+  = NoDiagnostics
+  | TimeAndLocation
+  | TimeLocationAndStats
+  deriving stock (Show)
+
+diagnosticLevelToCInt :: DiagnosticLevel -> CInt
+diagnosticLevelToCInt = \case
+  NoDiagnostics -> 0
+  TimeAndLocation -> 1
+  TimeLocationAndStats -> 2
+
+data CouldNotControl
+  = CouldNotControl Control CallStack
+  deriving anyclass (Exception)
+
+instance Show CouldNotControl where
+  show (CouldNotControl x c) =
+    mconcat
+      [ "Could not perform simulator control command: "
+      , show x
+      , "\n"
+      , prettyCallStack c
+      ]
+
+foreign import ccall "vpi_user.h vpi_control"
+  c_vpi_control_end :: CInt -> CInt -> IO Bool
+
+foreign import ccall "vpi_user.h vpi_control"
+  c_vpi_control_restart :: CInt -> CInt -> CInt -> CInt -> IO Bool
+
+controlSimulator :: forall o. Control -> SimCont o ()
+controlSimulator control =
+  liftOrThrow $
+    case control of
+      Stop d ->
+        c_vpi_control_end 66 (diagnosticLevelToCInt d)
```

Comment:
You might like c2hs for the constants here (66, 67, ...). You can automatically marshal `vpiStop`... etc. as a Haskell sum type.

https://github.com/vmchale/libarchive/blob/master/src/Codec/Archive/Types/Foreign.chs#L54

A lot of these constants are hard-coded and seem fickle! Unless hsc2hs/c2hs choke on the header files, I think they'd make this more robust + easier on your end.

Reply:
I didn't use `c2hs` because I wanted more control over failures to read. For some function `decode :: CInt -> MyType` with `c2hs` an input with no output means you get an `ErrorCall`, but in `clash-ffi` operations fail with more specific exception types. The reason being that sometimes errors are unavoidable and depending on the particular error you may want to recover, and this is somewhat tedious if all your errors become `ErrorCall`

Reply:
This module is a bit rough anyway, I should have made a type for these constants like I did with `Time` / `TimeType` (another instance where you have some type, but also want to refer to just it's tag as a separate type)

Reply:
You can still use hsc2hs! Something like `#{const vpiStop}`?

Reply:
True, I could also just `#include <vpi_user.h>` and use the macros directly from Haskell. That's probably the easiest from where I am now, and the case expressions would read nicer, e.g.

```haskell
case foo of
  vpiThing -> HaskellThing
  ...

Reply:
So I gave this a try and it didn't work because just including the C file isn't allowed when you're not using `c2hs` or something similar. I think `c2hs` constants might be nice if it works, or `c2hs` enum if it's possible to also use the pre-processor around those definitions (since the enum members depend on things like the Verilog spec version etc.)

I'm not sure if I want to rewrite things to use `c2hs` at this point, since it seems like a fair bit of work to change all the types / decoding functions across... I guess that's a debate for when I move this out of draft